File: benchmark/probprog/turing/standard/robust_regression.jl
# Turing.jl model spec for Robust Regression
# Matches NumPyro robust_regression.py and Impulse standard/robust_regression.jl

using Turing, Distributions, LinearAlgebra

@model function turing_robust_regression(X, Y, n, k, alpha_scale, beta_scale, beta_loc, sigma_mean)
    alpha ~ Normal(0.0, alpha_scale)
    beta ~ filldist(Normal(beta_loc, beta_scale), k)
    nu ~ Gamma(2.0, 1.0 / 0.1)  # Distributions.jl: Gamma(shape, scale); NumPyro: Gamma(conc, rate) → scale = 1/rate
    sigma ~ Exponential(sigma_mean)  # Distributions.jl: Exponential(scale); NumPyro: Exponential(rate) → scale = 1/rate
    mu = alpha .+ X * beta
    Turing.@addlogprob! sum(logpdf.(LocationScale.(mu, sigma, TDist(nu)), Y))
end

function setup(data)
    attrs = data["attrs"]
    n = Int(attrs["n"])
    k = Int(attrs["k"])
    alpha_scale = Float64(attrs["alpha_scale"])
    beta_scale = Float64(attrs["beta_scale"])
    beta_loc = Float64(attrs["beta_loc"])
    sigma_mean = Float64(attrs["sigma_mean"])

    X_raw = data["X"]
    X = zeros(Float64, n, k)
    for i in 1:n, j in 1:k
        X[i, j] = Float64(X_raw[i][j])
    end
    Y = Float64.(data["Y"])

    model = turing_robust_regression(X, Y, n, k, alpha_scale, beta_scale, beta_loc, sigma_mean)

    return (
        turing_model = model,
        model_name = "Robust Regression",
    )
end

function extract_samples(chain, num_samples)
    total = length(chain)
    start = total - num_samples + 1

    alpha = Array(chain[:alpha])[start:total]
    k = length([s for s in names(chain, :parameters) if startswith(string(s), "beta[")])
    beta = hcat([Array(chain[Symbol("beta[$i]")])[start:total] for i in 1:k]...)
    nu = Array(chain[:nu])[start:total]
    sigma = Array(chain[:sigma])[start:total]

    return Dict{String,Any}(
        "alpha" => collect(alpha),
        "beta" => [collect(beta[i, :]) for i in axes(beta, 1)],
        "nu" => collect(nu),
        "sigma" => collect(sigma),
    )
end

function get_init_params(data, init_params)
    if init_params === nothing
        return nothing
    end
    alpha = Float64(init_params["alpha"][1])
    beta = Float64.(init_params["beta"])
    nu = Float64(init_params["nu"][1])
    sigma = Float64(init_params["sigma"][1])
    return (alpha=alpha, beta=beta, nu=nu, sigma=sigma)
end

File: benchmark/probprog/pplbench/pplbench/ppls/pymc/robust_regression.py
from typing import Dict

import numpy as np
import pymc as pm
import xarray as xr

from .base_pymc_impl import BasePyMCImplementation


class RobustRegression(BasePyMCImplementation):
    def __init__(
        self,
        n: int,
        k: int,
        alpha_scale: float,
        beta_scale: float,
        beta_loc: float,
        sigma_mean: float,
    ) -> None:
        self.n = n
        self.k = k
        self.alpha_scale = alpha_scale
        self.beta_scale = beta_scale
        self.beta_loc = beta_loc
        self.sigma_mean = sigma_mean

    def build_model(self, data: xr.Dataset):
        data = data.transpose("item", "feature")
        X, Y = data.X.values, data.Y.values

        with pm.Model() as model:
            alpha = pm.Normal("alpha", mu=0.0, sigma=self.alpha_scale)
            beta = pm.Normal("beta", mu=self.beta_loc, sigma=self.beta_scale,
                             shape=self.k)
            # Gamma(shape=2, rate=0.1) -> PyMC: Gamma(alpha=2, beta=0.1)
            nu = pm.Gamma("nu", alpha=2.0, beta=0.1)
            # Exponential(rate=1/sigma_mean) -> PyMC uses 1/lam as scale
            sigma = pm.Exponential("sigma", lam=1.0 / self.sigma_mean)
            mu = alpha + X @ beta
            pm.StudentT("Y", nu=nu, mu=mu, sigma=sigma, observed=Y)

        return model

    def extract_data_from_pymc(
        self, samples: Dict[str, np.ndarray]
    ) -> xr.Dataset:
        return xr.Dataset(
            {
                "alpha": (["draw"], samples["alpha"]),
                "beta": (["draw", "feature"], samples["beta"]),
                "nu": (["draw"], samples["nu"]),
                "sigma": (["draw"], samples["sigma"]),
            },
            coords={
                "draw": np.arange(samples["beta"].shape[0]),
                "feature": np.arange(samples["beta"].shape[1]),
            },
        )
